server/src/main/java/org/opensearch/client/node/PluginNodeClient.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.client.node;
+
+import org.opensearch.action.ActionRequest;
+import org.opensearch.action.ActionType;
+import org.opensearch.client.Client;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.util.concurrent.PluginContextSwitcher;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.action.ActionResponse;
+import org.opensearch.plugins.Plugin;
+import org.opensearch.tasks.Task;
+import org.opensearch.tasks.TaskListener;
+import org.opensearch.threadpool.ThreadPool;
+
+/**
+ * Client that executes actions on the local node with contextual information about the plugin executing
+ * transport actions.
+ *
+ * @opensearch.api
+ */
+public class PluginNodeClient extends NodeClient {
```

Comment:
What happens when a plugin uses their own threading construct to run both the node client and plugin node client at the same time?  I think there are 3 possibilities:

1. Both clients reuse the same thread context, depending on order of execution some requests are using the expected plugin context, some are using the context of the request.
2. Both clients always run in their own isolated thread context, `switchContext()` never fails to correctly set the context.
3. We do not allow the plugin node client to run actions locally, they all are routed through the transport layer where the context is serialized.

If I had to bet, I would say that we are sitting on possibility 1.  Even if we are lucky and in possibility 2 space, I think we need to verify this is the case.  Maybe we could work backwards from the test scenario and then figure out if we need to adjust how the plugin client / context switcher works, what do you think?

Reply:
In addition to those questions, what prevents any plugin from creating `NodeClient` directly?

```
NodeClient client = new NodeClient(settings, threadPool);
```

Reply:
The issue isn't the nodeClient though. Running `nodeClient.execute(TransportAction...)` runs the action in the context of the authenticated user when the security plugin is installed. The issue is with `ThreadContext.stashContext()` which will temporarily remove info about the authenticated user from the ThreadContext which allows a plugin to run transport actions w/o authz checks.

FYI NodeClient must be initialized like [this](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/node/Node.java#L1449-L1454) before it can be used to execute transport actions.

Reply:
> The issue isn't the nodeClient though. Running `nodeClient.execute(TransportAction...)` runs the action in the context of the authenticated user when the security plugin is installed.

The mechanism of caller identity impersonation are suggested here https://github.com/opensearch-project/OpenSearch/pull/14630#discussion_r1713676514. I don't see the reasons why existing `NodeClient` would not benefit from that?

Reply:
@peternied I just confirmed with a test. There is an existing issue with multi-threading in plugins. If a plugin spawns a new thread, the child thread does not have the same context as its parent thread. 

From what I can tell, all plugins have the ability to spawn threads. If a plugin wants to do multi-threading they currently need to keep a copy of the storedContext and restore it in the child thread.

There is an example of this in the alerting plugin [here](https://github.com/opensearch-project/alerting/blob/07fb52227c6641d61cc08faeb4426a1767d5e805/alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetRemoteIndexesAction.kt#L99-L104).

Assuming that handling an API request is done in a single thread, then a transport action is either executed in the context of the authenticated user or in the plugin system context. It cannot do both simultaneously.

Reply:
Based on these [entries in the security.policy](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/resources/org/opensearch/bootstrap/security.policy#L147-L150) file. Every codebase in the project is given permission by default to create new threads. It is possible to move those permissions to the [core-only](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/resources/org/opensearch/bootstrap/security.policy#L44-L53) section of the policy file, but there are probably many plugins that rely on that and I'd be reluctant to make that change. I wonder if its possible to make child threads inherit the context of the parent by default or enforce a certain pattern of spawning a new thread within plugins.